File: src/api/models/case.py
"""
Dispute Case data models.

Mirrors src/shared/schemas/case.schema.json (JSON Schema draft 2020-12).
Hand-maintained for now; see src/shared/README.md for the contract.
Update this file whenever the JSON Schema changes.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal, Optional


# ── String-literal unions (mirror $defs in the JSON Schema) ──────────────────

CaseStatus = Literal[
    "intake",
    "evidence_gathering",
    "ai_drafting",
    "pending_review",
    "approved",
    "denied",
    "escalated",
    "submitted",
    "expired",
]

CardNetwork = Literal["visa", "mastercard", "amex", "discover"]

RiskLevel = Literal["low", "medium", "high", "critical"]

EvidenceType = Literal[
    "transaction",
    "shipping",
    "communication",
    "receipt",
    "contract",
    "fraud_signal",
    "order",
]

CompletenessLevel = Literal["complete", "partial", "missing"]

ImpactLevel = Literal["critical", "high", "medium", "low"]


# ── Nested object models ──────────────────────────────────────────────────────

@dataclass
class ReasonCodeChecklistItem:
    item: str
    required: bool
    satisfied: bool


@dataclass
class Evidence:
    evidenceId: str            # UUID
    type: EvidenceType
    sourceSystem: str
    retrievedAt: str           # ISO 8601 date-time
    contentRef: str            # Blob URI or document ID
    completeness: CompletenessLevel


@dataclass
class EvidenceGap:
    missingItem: str
    reason: str
    impact: ImpactLevel


@dataclass
class Citation:
    evidenceId: str
    excerpt: str


@dataclass
class RebuttalDraft:
    text: str = ""
    citations: list[Citation] = field(default_factory=list)


@dataclass
class Deadline:
    network: str
    dueDate: str               # ISO 8601 date
    daysRemaining: int


@dataclass
class CaseSummaryDeadline:
    """Deadline subset for CaseSummary — omits network."""
    dueDate: str               # ISO 8601 date
    daysRemaining: int


# ── Top-level models ──────────────────────────────────────────────────────────

@dataclass
class Case:
    """Full dispute case record."""
    # Required fields
    caseId: str                # UUID
    status: CaseStatus
    reasonCode: str            # e.g. "Visa 13.1"
    deadline: Deadline
    createdAt: str             # ISO 8601 date-time

    # Identifiers (optional)
    orchestrationId: Optional[str] = None      # Durable Functions instance ID (equals caseId)
    disputeRef: Optional[str] = None           # Network ARN / reference
    cardNetwork: Optional[CardNetwork] = None
    merchantName: Optional[str] = None
    cardholderName: Optional[str] = None
    caseDescription: Optional[str] = None
    transactionAmount: Optional[float] = None
    transactionDate: Optional[str] = None      # ISO 8601 date

    # Reason code detail
    reasonCodeLabel: Optional[str] = None
    reasonCodeChecklist: list[ReasonCodeChecklistItem] = field(default_factory=list)

    # Evidence
    evidence: list[Evidence] = field(default_factory=list)
    evidenceGaps: list[EvidenceGap] = field(default_factory=list)

    # Scoring
    winProbability: Optional[float] = None     # 0–1
    riskLevel: Optional[RiskLevel] = None

    # Rebuttal
    rebuttalDraft: Optional[RebuttalDraft] = None

    # Timestamps
    updatedAt: Optional[str] = None            # ISO 8601 date-time
    resolvedAt: Optional[str] = None           # ISO 8601 date-time


@dataclass
class CaseSummary:
    """
    Lightweight queue-list subset.
    Mirrors CaseSummary in the JSON Schema — returned by GET /cases.
    """
    # Required fields
    caseId: str
    status: CaseStatus
    reasonCode: str
    deadline: CaseSummaryDeadline
    createdAt: str             # ISO 8601 date-time

    # Optional fields
    cardNetwork: Optional[CardNetwork] = None
    merchantName: Optional[str] = None
    caseDescription: Optional[str] = None
    transactionAmount: Optional[float] = None
    reasonCodeLabel: Optional[str] = None
    winProbability: Optional[float] = None     # 0–1
    riskLevel: Optional[RiskLevel] = None
    updatedAt: Optional[str] = None            # ISO 8601 date-time

File: src/web/src/types/case.ts
/**
 * Dispute Case TypeScript types.
 *
 * Mirrors src/shared/schemas/case.schema.json (JSON Schema draft 2020-12).
 * Hand-maintained for now; see src/shared/README.md for the contract.
 * Update this file whenever the JSON Schema changes.
 */

// ── String-literal unions (mirror $defs in the JSON Schema) ──────────────────

export type CaseStatus =
  | "intake"
  | "evidence_gathering"
  | "ai_drafting"
  | "pending_review"
  | "approved"
  | "denied"
  | "escalated"
  | "submitted"
  | "expired"
  | "closed";

export type CardNetwork = "visa" | "mastercard" | "amex" | "discover";

export type RiskLevel = "low" | "medium" | "high" | "critical";

export type EvidenceType =
  | "transaction"
  | "shipping"
  | "communication"
  | "receipt"
  | "contract"
  | "fraud_signal"
  | "order"
  | "photo"
  | "fraud_screening"
  | "device_fingerprint";

export type CompletenessLevel = "complete" | "partial" | "missing";

export type ImpactLevel = "critical" | "high" | "medium" | "low";

// ── Nested object interfaces ──────────────────────────────────────────────────

export interface ReasonCodeChecklistItem {
  item: string;
  required: boolean;
  satisfied: boolean;
}

export interface Evidence {
  evidenceId: string;       // UUID
  type: EvidenceType;
  sourceSystem: string;
  retrievedAt: string;      // ISO 8601 date-time
  contentRef: string;       // Blob URI or document ID
  completeness: CompletenessLevel;
}

export interface EvidenceGap {
  missingItem: string;
  reason: string;
  impact: ImpactLevel;
  suggestedAction?: string;
}

export interface Citation {
  evidenceId: string;
  excerpt: string;
}

export interface RebuttalDraft {
  text: string;
  citations: Citation[];
}

export interface Deadline {
  network: string;
  dueDate: string;          // ISO 8601 date
  daysRemaining: number;
}

/** Deadline subset for CaseSummary — omits network. */
export interface CaseSummaryDeadline {
  dueDate: string;          // ISO 8601 date
  daysRemaining: number;
}

// ── Timeline types ────────────────────────────────────────────────────────────

export type TimelineEventType =
  | "case_created"
  | "evidence_retrieved"
  | "evidence_gap_detected"
  | "score_generated"
  | "ai_draft_generated"
  | "status_changed"
  | "status_change"
  | "analyst_assigned"
  | "analyst_action"
  | "escalated"
  | "deadline_warning"
  | "document_uploaded"
  | "comment_added"
  | "analyst_note"
  | "ai_recommendation_response"
  | "evidence_gap_requested"
  | "customer_response"
  | "case_closed_artifact_created"
  | "customer_response_requested"
  | "customer_response_received"
  // Transaction events
  | "authorization"
  | "clearing"
  | "settlement"
  | "refund"
  | "chargeback"
  | "representment"
  | "arbitration"
  // Customer activity
  | "customer_login"
  | "device_change"
  | "mfa_event"
  | "geolocation"
  | "previous_purchase"
  // Fraud signals
  | "velocity_alert"
  | "device_reputation"
  | "ip_reputation"
  | "high_risk_indicator"
  | "friendly_fraud_indicator";

export type TimelineCategory = "transaction" | "customer_activity" | "fraud_signals" | "case_activity";

export interface TimelineEvent {
  eventId: string;
  disputeId: string;
  eventType: TimelineEventType;
  timestamp: string;         // ISO 8601 date-time
  actor?: string;            // system | analyst name
  description: string;
  metadata?: Record<string, unknown>;
}

// ── Top-level interfaces ──────────────────────────────────────────────────────

/** Full dispute case record. */
export interface Case {
  // Required
  caseId: string;                        // UUID
  status: CaseStatus;
  reasonCode: string;                    // e.g. "Visa 13.1"
  deadline: Deadline;
  createdAt: string;                     // ISO 8601 date-time

  // Assignment (optional — null/undefined = unassigned = "Open")
  assignedAnalystId?: string;
  assignedAnalystName?: string;

  // Identifiers (optional)
  orchestrationId?: string;              // Durable Functions instance ID (equals caseId)
  disputeRef?: string;                   // Network ARN / reference
  cardNetwork?: CardNetwork;
  merchantName?: string;
  cardholderName?: string;
  caseDescription?: string;
  transactionAmount?: number;
  transactionDate?: string;              // ISO 8601 date

  // Reason code detail (optional)
  reasonCodeLabel?: string;
  reasonCodeChecklist?: ReasonCodeChecklistItem[];

  // Evidence (optional)
  evidence?: Evidence[];
  evidenceGaps?: EvidenceGap[];

  // Scoring (optional)
  winProbability?: number;               // 0–1
  riskLevel?: RiskLevel;

  // Rebuttal (optional)
  rebuttalDraft?: RebuttalDraft;

  // Timestamps (optional)
  updatedAt?: string;                    // ISO 8601 date-time
  resolvedAt?: string;                   // ISO 8601 date-time
}

/**
 * Lightweight queue-list subset.
 * Mirrors CaseSummary in the JSON Schema — returned by GET /cases.
 */
export interface CaseSummary {
  // Required
  caseId: string;
  status: CaseStatus;
  reasonCode: string;
  deadline: CaseSummaryDeadline;
  createdAt: string;                     // ISO 8601 date-time

  // Assignment (optional — null/undefined = unassigned = "Open")
  assignedAnalystId?: string;
  assignedAnalystName?: string;

  // Optional
  cardNetwork?: CardNetwork;
  merchantName?: string;
  caseDescription?: string;
  transactionAmount?: number;
  reasonCodeLabel?: string;
  winProbability?: number;               // 0–1
  riskLevel?: RiskLevel;
  updatedAt?: string;                    // ISO 8601 date-time
  lastActivityAt?: string;               // ISO 8601 date-time
  lastActivityType?: string;
  lastActivityActor?: string;
  lastActivityDetail?: string;
}
